From 1eaf6f71309e95b8a4cdea2d88969266685b3c10 Mon Sep 17 00:00:00 2001
From: Chen Hanxiao <chenhanxiao@cn.fujitsu.com>
Date: Wed, 22 Jan 2014 17:44:38 +0800
Subject: [PATCH 2/2] virt-install: add support for '--panic option'

Signed-off-by: Chen Hanxiao <chenhanxiao@cn.fujitsu.com>

(crobinso: man page and cli tweaks)

Closes: https://bugzilla.redhat.com/show_bug.cgi?id=1101536

(cherry picked from commit a98515a4da626cd5e7b4e8d78a729e6f071c6376)

Conflicts:
	man/en/virt-install.pod
	virt-install
	virtinst/cli.py
---
 man/en/virt-install.pod    |  8 ++++++++
 man/en/virt-install.pod.in |  8 ++++++++
 virt-install               |  1 +
 virtinst/cli.py            | 40 ++++++++++++++++++++++++++++++++++++++++
 4 files changed, 57 insertions(+)

diff --git a/man/en/virt-install.pod b/man/en/virt-install.pod
index 7c4b64d..e5f5001 100644
--- a/man/en/virt-install.pod
+++ b/man/en/virt-install.pod
@@ -1163,6 +1163,14 @@ Add a USB device redirected via a dedicated Spice channel.
 
 =back
 
+=item --panic OPTS
+
+Attach a panic notifier device to the guest. For the recommended settings, use:
+
+--panic default
+
+See C<http://libvirt.org/formatdomain.html#elementsPanic> for complete details.
+
 =back
 
 =head2 Miscellaneous Options
diff --git a/man/en/virt-install.pod.in b/man/en/virt-install.pod.in
index c21288c..cf21c71 100644
--- a/man/en/virt-install.pod.in
+++ b/man/en/virt-install.pod.in
@@ -1094,6 +1094,14 @@ Add a USB device redirected via a dedicated Spice channel.
 
 =back
 
+=item --panic OPTS
+
+Attach a panic notifier device to the guest. For the recommended settings, use:
+
+--panic default
+
+See C<http://libvirt.org/formatdomain.html#elementsPanic> for complete details.
+
 =back
 
 =head2 Miscellaneous Options
diff --git a/virt-install b/virt-install
index 975093a..f094c4f 100755
--- a/virt-install
+++ b/virt-install
@@ -506,6 +506,7 @@ def build_guest_instance(conn, options):
                  guest, cli.parse_console)
     cli.get_hostdevs(options.hostdevs, guest)
     cli.get_smartcard(guest, options.smartcard)
+    cli.get_panic(guest, options.panic)
 
 
     # Install options
diff --git a/virtinst/cli.py b/virtinst/cli.py
index e9b6e3b..5a27017 100644
--- a/virtinst/cli.py
+++ b/virtinst/cli.py
@@ -1100,6 +1100,9 @@ def add_device_options(devg):
     devg.add_option("", "--redirdev", dest="redirdev", action="append",
                     help=_("Configure a guest redirection device. Ex:\n"
                            "--redirdev usb,type=tcp,server=192.168.1.1:4000"))
+    devg.add_option("--panic", dest="panic", action="append",
+                    help=_("Configure a guest panic device. Ex:\n"
+                           "--panic default"))
 
 def add_gfx_option(devg):
     devg.add_option("", "--graphics", dest="graphics", action="append",
@@ -1795,6 +1798,43 @@ def parse_watchdog(guest, optstring, dev=None):
     return dev
 
 
+###################
+# --panic parsing #
+###################
+
+def parse_panic(guest, optstr, dev=None):
+    ignore = guest
+
+    if not dev:
+        dev = virtinst.VirtualPanicDevice(guest.conn)
+
+    if optstr == "default":
+        return dev
+
+    opts = parse_optstr(optstr)
+    set_param = _build_set_param(dev, opts)
+
+    iobase = get_opt_param(opts, "iobase")
+    type = get_opt_param(opts, "type")
+
+    if iobase:
+        dev.iobase = iobase
+    if type:
+        dev.type = type
+
+    return dev
+
+def get_panic(guest, sc_opts):
+    for sc in listify(sc_opts):
+        try:
+            dev = parse_panic(guest, sc)
+        except Exception, e:
+            fail(_("Error in pvpanic device parameters: %s") % str(e))
+
+        if dev:
+            guest.add_device(dev)
+
+
 ######################################################
 # --serial, --parallel, --channel, --console parsing #
 ######################################################
-- 
1.9.3

